packages/block-library/src/gallery/edit.js

```diff
@@ -105,11 +105,22 @@ class GalleryEdit extends Component {
 		};
 	}
 
-	onSelectImages( images ) {
-		const { columns } = this.props.attributes;
+	onSelectImages( newImages ) {
+		const { columns, images } = this.props.attributes;
+
 		this.setAttributes( {
-			images: images.map( ( image ) => pickRelevantMediaFiles( image ) ),
-			columns: columns ? Math.min( images.length, columns ) : columns,
+			images: newImages.map( ( image ) => {
+				const newImage = pickRelevantMediaFiles( image );
+				let oldImage = filter( images, { id: newImage.id } );
+				if ( oldImage.length > 0 ) {
+					oldImage = oldImage.reduce( ( img ) => img );
+					if ( oldImage.caption !== '' ) {
+						newImage.caption = oldImage.caption;
+					}
+				}
+				return newImage;
+			} ),
```

Comment:
We could use [`_.groupBy`](https://devdocs.io/lodash~4/index#groupBy) and [`_.get`](https://devdocs.io/lodash~4/index#get) to make this run a little faster and read more declaratively. 

```js
const { columns, images } = this.props.attributes;

const imagesById = groupBy( images, 'id' );

this.setAttributes( {
	images: newImages.map( ( image ) => ( {
		...pickRelevantMediaFiles( image ),
		caption: get( imageById, [ image.id, 'caption' ], image.caption ),
	} ),
	columns: columns ? Math.min( newImages.length, columns ) : columns,
} );

Reply:
yep, updated :)